src/nf/nf_dropout_layer_submodule.f90

```diff
@@ -0,0 +1,78 @@
+submodule (nf_dropout_layer) nf_dropout_layer_submodule
+  use nf_random, only: shuffle
+  !! This submodule implements the procedures defined in the
+  !! nf_dropout_layer module.
+
+contains
+
+  module function dropout_layer_cons(rate) result(res)
+    real, intent(in) :: rate
+    type(dropout_layer) :: res
+    res % dropout_rate = rate
+  end function dropout_layer_cons
+
+
+  module subroutine init(self, input_shape)
+    class(dropout_layer), intent(in out) :: self
+    integer, intent(in) :: input_shape(:)
+
+    self % input_size = input_shape(1)
+
+    ! Allocate arrays
+    allocate(self % output(self % input_size))
+    allocate(self % gradient(self % input_size))
+    allocate(self % mask(self % input_size))
+
+    ! Initialize arrays
+    self % output = 0
+    self % gradient = 0
+    self % mask = 1  ! Default mask is all ones (no dropout)
+
+  end subroutine init
+
+
+  module subroutine forward(self, input)
+    class(dropout_layer), intent(in out) :: self
+    real, intent(in) :: input(:)
+
+    ! Generate random mask for dropout, training mode only
+    if (self % training) then
+
+      ! Set the first dropout_rate number of elements to 0, the rest to 1,
+      ! and shuffle. Note that the selection of the elements rounds down to
+      ! the nearest integer, so in cases where size(input) * dropout_rate is
+      ! not an integer, the actual dropout rate will be slightly lower.
+      self % mask = 1
+      self % mask(:int(size(self % mask) * self % dropout_rate)) = 0
+      call shuffle(self % mask)
+
+      ! Scale factor to preserve the input sum
+      self % scale = 1 / (1 - self % dropout_rate)
+
+      ! Apply dropout mask
+      self % output = input * self % mask * self % scale
```

Comment:
I wonder what would be the most efficient. For example, if `mask` is `logical` the following could be done:
```fortran
self % output = merge(input * self%scale, 0., self%mask)
```

Reply:
Good idea, it's not obvious to me either, I'll do a few tests to see if there's any difference.

Reply:
In some very simple timing tests, using `integer` mask was indistinguishable from `real`. Using `merge` with `logical` mask appears a few percent slower based on ~10 repeats. Will keep it as is.